cpp/include/raft/core/memory_type.hpp

```diff
@@ -14,13 +14,28 @@
  * limitations under the License.
  */
 #pragma once
+#include <cstdint>
+#include <optional>
+#ifndef RAFT_DISABLE_CUDA
+#include <cuda_runtime.h>
+#include <raft/util/cuda_rt_essentials.hpp>
+#include <type_traits>
+#else
+#include <raft/core/logger.hpp>
+#endif
 
 namespace raft {
-enum class memory_type { host, device, managed, pinned };
+enum class memory_type : std::uint8_t {
```

Comment:
nit: I sorta wonder if we should align this enum with the cudaMemoryType enum:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1g13de56a8fe75569530ecc3a3106e9b6d

afaict the cudaMemoryType enum has the exact same values (aside from some naming differences with pinned memory, like cudaMemoryTypeUnregistered being unpinned host memory).

Reply:
Agreed! Time to see how well I did at ensuring that the order of the enum does not matter ;). I'm kicking off a build now with the enum values aligned.

A general note for RAFT's memory_type enum is that it was designed to be extensible beyond the usual CUDA memory types. For instance, we might choose to add a `cuda_ipc` memory type in the future that allows us to easily flag memory that can be shared across processes. That does not mean that the base values shouldn't currently align with cudaMemoryType; just that we may need to be careful if we do decide to extend these values in the future.